File: crypto_agent/backtest.py
from dataclasses import dataclass, field

from .indicators import compute_sma, compute_rsi, compute_bollinger


@dataclass
class BacktestResult:
    strategy: str
    symbol: str
    timeframe: str
    total_return: float
    max_drawdown: float
    sharpe_ratio: float
    win_rate: float
    total_trades: int
    trades: list[dict] = field(default_factory=list)
    equity_curve: list[float] = field(default_factory=list)


class BacktestEngine:
    def __init__(self, initial_capital: float = 10000.0):
        self.initial_capital = initial_capital

    def run(
        self,
        ohlcv: list[dict],
        strategy: str,
        params: dict | None = None,
        symbol: str = "",
        timeframe: str = "",
    ) -> BacktestResult:
        params = params or {}
        strategies = {
            "sma_crossover": self._run_sma_crossover,
            "rsi_reversal": self._run_rsi_reversal,
            "bollinger_bounce": self._run_bollinger_bounce,
        }
        if strategy not in strategies:
            raise ValueError(
                f"Unknown strategy: {strategy}. Available: {list(strategies.keys())}"
            )
        trades, equity = strategies[strategy](ohlcv, params)
        return self._build_result(strategy, symbol, timeframe, trades, equity)

    # -- Strategy runners --------------------------------------------------

    def _run_sma_crossover(
        self, ohlcv: list[dict], params: dict
    ) -> tuple[list[dict], list[float]]:
        short_period = params.get("short_period", 10)
        long_period = params.get("long_period", 30)
        closes = [bar["close"] for bar in ohlcv]

        cash = self.initial_capital
        position = 0.0
        trades: list[dict] = []
        equity: list[float] = []

        for i in range(long_period, len(closes)):
            price = closes[i]
            window = closes[: i + 1]
            short_sma = compute_sma(window, short_period)
            long_sma = compute_sma(window, long_period)

            if short_sma > long_sma and position == 0:
                qty = cash / price
                position = qty
                cash = 0.0
                trades.append(
                    {"side": "buy", "price": price, "quantity": qty, "index": i}
                )
            elif short_sma < long_sma and position > 0:
                cash = position * price
                trades.append(
                    {"side": "sell", "price": price, "quantity": position, "index": i}
                )
                position = 0.0

            equity.append(cash + position * price)

        self._close_open_position(closes, position, cash, trades, equity)
        return trades, equity

    def _run_rsi_reversal(
        self, ohlcv: list[dict], params: dict
    ) -> tuple[list[dict], list[float]]:
        period = params.get("period", 14)
        oversold = params.get("oversold", 30)
        overbought = params.get("overbought", 70)
        closes = [bar["close"] for bar in ohlcv]

        cash = self.initial_capital
        position = 0.0
        trades: list[dict] = []
        equity: list[float] = []

        for i in range(period + 1, len(closes)):
            price = closes[i]
            window = closes[: i + 1]
            rsi = compute_rsi(window, period)

            if rsi < oversold and position == 0:
                qty = cash / price
                position = qty
                cash = 0.0
                trades.append(
                    {"side": "buy", "price": price, "quantity": qty, "index": i}
                )
            elif rsi > overbought and position > 0:
                cash = position * price
                trades.append(
                    {"side": "sell", "price": price, "quantity": position, "index": i}
                )
                position = 0.0

            equity.append(cash + position * price)

        self._close_open_position(closes, position, cash, trades, equity)
        return trades, equity

    def _run_bollinger_bounce(
        self, ohlcv: list[dict], params: dict
    ) -> tuple[list[dict], list[float]]:
        period = params.get("period", 20)
        std_dev = params.get("std_dev", 2.0)
        closes = [bar["close"] for bar in ohlcv]

        cash = self.initial_capital
        position = 0.0
        trades: list[dict] = []
        equity: list[float] = []

        for i in range(period, len(closes)):
            price = closes[i]
            window = closes[: i + 1]
            lower, _mid, upper = compute_bollinger(window, period, std_dev)

            if price < lower and position == 0:
                qty = cash / price
                position = qty
                cash = 0.0
                trades.append(
                    {"side": "buy", "price": price, "quantity": qty, "index": i}
                )
            elif price > upper and position > 0:
                cash = position * price
                trades.append(
                    {"side": "sell", "price": price, "quantity": position, "index": i}
                )
                position = 0.0

            equity.append(cash + position * price)

        self._close_open_position(closes, position, cash, trades, equity)
        return trades, equity

    # -- Helpers -----------------------------------------------------------

    @staticmethod
    def _close_open_position(
        closes: list[float],
        position: float,
        cash: float,
        trades: list[dict],
        equity: list[float],
    ) -> None:
        if position > 0 and closes:
            final_price = closes[-1]
            cash = position * final_price
            trades.append(
                {
                    "side": "sell",
                    "price": final_price,
                    "quantity": position,
                    "index": len(closes) - 1,
                }
            )
            if equity:
                equity[-1] = cash

    def _build_result(
        self,
        strategy: str,
        symbol: str,
        timeframe: str,
        trades: list[dict],
        equity: list[float],
    ) -> BacktestResult:
        if equity:
            total_return = (
                (equity[-1] - self.initial_capital) / self.initial_capital
            ) * 100
        else:
            total_return = 0.0

        return BacktestResult(
            strategy=strategy,
            symbol=symbol,
            timeframe=timeframe,
            total_return=total_return,
            max_drawdown=self._calc_max_drawdown(equity),
            sharpe_ratio=self._calc_sharpe(equity),
            win_rate=self._calc_win_rate(trades),
            total_trades=len(trades),
            trades=trades,
            equity_curve=equity,
        )

    @staticmethod
    def _calc_max_drawdown(equity: list[float]) -> float:
        if not equity:
            return 0.0
        peak = equity[0]
        max_dd = 0.0
        for value in equity:
            if value > peak:
                peak = value
            dd = (peak - value) / peak * 100
            if dd > max_dd:
                max_dd = dd
        return max_dd

    @staticmethod
    def _calc_sharpe(equity: list[float]) -> float:
        if len(equity) < 2:
            return 0.0
        returns = [
            (equity[i] - equity[i - 1]) / equity[i - 1]
            for i in range(1, len(equity))
            if equity[i - 1] != 0
        ]
        if len(returns) < 2:
            return 0.0
        avg_ret = sum(returns) / len(returns)
        variance = sum((r - avg_ret) ** 2 for r in returns) / (len(returns) - 1)
        std_ret = variance**0.5
        if std_ret == 0:
            return 0.0
        return (avg_ret / std_ret) * (252**0.5)

    @staticmethod
    def _calc_win_rate(trades: list[dict]) -> float:
        wins = 0
        total = 0
        for i in range(0, len(trades) - 1, 2):
            if trades[i]["side"] == "buy" and trades[i + 1]["side"] == "sell":
                if trades[i + 1]["price"] > trades[i]["price"]:
                    wins += 1
                total += 1
        return (wins / total * 100) if total > 0 else 0.0

File: src/backtest.ts
import { computeSma, computeRsi, computeBollinger } from "./indicators.js";

export interface BacktestResult {
  strategy: string;
  symbol: string;
  timeframe: string;
  totalReturn: number;
  maxDrawdown: number;
  sharpeRatio: number;
  winRate: number;
  totalTrades: number;
  trades: Array<Record<string, any>>;
  equityCurve: number[];
}

interface OhlcvBar {
  close: number;
  [key: string]: any;
}

export class BacktestEngine {
  private initialCapital: number;

  constructor(initialCapital = 10000) {
    this.initialCapital = initialCapital;
  }

  run(
    ohlcv: OhlcvBar[],
    strategy: string,
    params?: Record<string, any> | null,
    symbol = "",
    timeframe = "",
  ): BacktestResult {
    const p = params ?? {};
    const strategies: Record<string, (o: OhlcvBar[], p: Record<string, any>) => [any[], number[]]> = {
      sma_crossover: (o, p) => this.runSmaCrossover(o, p),
      rsi_reversal: (o, p) => this.runRsiReversal(o, p),
      bollinger_bounce: (o, p) => this.runBollingerBounce(o, p),
    };
    if (!(strategy in strategies)) {
      throw new Error(`Unknown strategy: ${strategy}. Available: ${Object.keys(strategies).join(", ")}`);
    }
    const [trades, equity] = strategies[strategy](ohlcv, p);
    return this.buildResult(strategy, symbol, timeframe, trades, equity);
  }

  private runSmaCrossover(ohlcv: OhlcvBar[], params: Record<string, any>): [any[], number[]] {
    const shortPeriod = params.short_period ?? 10;
    const longPeriod = params.long_period ?? 30;
    const closes = ohlcv.map((b) => b.close);
    let cash = this.initialCapital;
    let position = 0;
    const trades: any[] = [];
    const equity: number[] = [];

    for (let i = longPeriod; i < closes.length; i++) {
      const price = closes[i];
      const window = closes.slice(0, i + 1);
      const shortSma = computeSma(window, shortPeriod);
      const longSma = computeSma(window, longPeriod);

      if (shortSma > longSma && position === 0) {
        const qty = cash / price;
        position = qty;
        cash = 0;
        trades.push({ side: "buy", price, quantity: qty, index: i });
      } else if (shortSma < longSma && position > 0) {
        cash = position * price;
        trades.push({ side: "sell", price, quantity: position, index: i });
        position = 0;
      }
      equity.push(cash + position * price);
    }
    this.closeOpenPosition(closes, position, cash, trades, equity);
    return [trades, equity];
  }

  private runRsiReversal(ohlcv: OhlcvBar[], params: Record<string, any>): [any[], number[]] {
    const period = params.period ?? 14;
    const oversold = params.oversold ?? 30;
    const overbought = params.overbought ?? 70;
    const closes = ohlcv.map((b) => b.close);
    let cash = this.initialCapital;
    let position = 0;
    const trades: any[] = [];
    const equity: number[] = [];

    for (let i = period + 1; i < closes.length; i++) {
      const price = closes[i];
      const window = closes.slice(0, i + 1);
      const rsi = computeRsi(window, period);

      if (rsi < oversold && position === 0) {
        const qty = cash / price;
        position = qty;
        cash = 0;
        trades.push({ side: "buy", price, quantity: qty, index: i });
      } else if (rsi > overbought && position > 0) {
        cash = position * price;
        trades.push({ side: "sell", price, quantity: position, index: i });
        position = 0;
      }
      equity.push(cash + position * price);
    }
    this.closeOpenPosition(closes, position, cash, trades, equity);
    return [trades, equity];
  }

  private runBollingerBounce(ohlcv: OhlcvBar[], params: Record<string, any>): [any[], number[]] {
    const period = params.period ?? 20;
    const stdDevParam = params.std_dev ?? 2;
    const closes = ohlcv.map((b) => b.close);
    let cash = this.initialCapital;
    let position = 0;
    const trades: any[] = [];
    const equity: number[] = [];

    for (let i = period; i < closes.length; i++) {
      const price = closes[i];
      const window = closes.slice(0, i + 1);
      const [lower, , upper] = computeBollinger(window, period, stdDevParam);

      if (price < lower && position === 0) {
        const qty = cash / price;
        position = qty;
        cash = 0;
        trades.push({ side: "buy", price, quantity: qty, index: i });
      } else if (price > upper && position > 0) {
        cash = position * price;
        trades.push({ side: "sell", price, quantity: position, index: i });
        position = 0;
      }
      equity.push(cash + position * price);
    }
    this.closeOpenPosition(closes, position, cash, trades, equity);
    return [trades, equity];
  }

  private closeOpenPosition(closes: number[], position: number, _cash: number, trades: any[], equity: number[]): void {
    if (position > 0 && closes.length) {
      const finalPrice = closes[closes.length - 1];
      const cash = position * finalPrice;
      trades.push({ side: "sell", price: finalPrice, quantity: position, index: closes.length - 1 });
      if (equity.length) equity[equity.length - 1] = cash;
    }
  }

  private buildResult(
    strategy: string,
    symbol: string,
    timeframe: string,
    trades: any[],
    equity: number[],
  ): BacktestResult {
    const totalReturn = equity.length
      ? ((equity[equity.length - 1] - this.initialCapital) / this.initialCapital) * 100
      : 0;
    return {
      strategy,
      symbol,
      timeframe,
      totalReturn,
      maxDrawdown: this.calcMaxDrawdown(equity),
      sharpeRatio: this.calcSharpe(equity),
      winRate: this.calcWinRate(trades),
      totalTrades: trades.length,
      trades,
      equityCurve: equity,
    };
  }

  private calcMaxDrawdown(equity: number[]): number {
    if (!equity.length) return 0;
    let peak = equity[0];
    let maxDd = 0;
    for (const value of equity) {
      if (value > peak) peak = value;
      const dd = ((peak - value) / peak) * 100;
      if (dd > maxDd) maxDd = dd;
    }
    return maxDd;
  }

  private calcSharpe(equity: number[]): number {
    if (equity.length < 2) return 0;
    const returns: number[] = [];
    for (let i = 1; i < equity.length; i++) {
      if (equity[i - 1] !== 0) returns.push((equity[i] - equity[i - 1]) / equity[i - 1]);
    }
    if (returns.length < 2) return 0;
    const avgRet = returns.reduce((a, b) => a + b, 0) / returns.length;
    const variance = returns.reduce((sum, r) => sum + (r - avgRet) ** 2, 0) / (returns.length - 1);
    const stdRet = Math.sqrt(variance);
    if (stdRet === 0) return 0;
    return (avgRet / stdRet) * Math.sqrt(252);
  }

  private calcWinRate(trades: any[]): number {
    let wins = 0;
    let total = 0;
    for (let i = 0; i < trades.length - 1; i += 2) {
      if (trades[i].side === "buy" && trades[i + 1].side === "sell") {
        if (trades[i + 1].price > trades[i].price) wins++;
        total++;
      }
    }
    return total > 0 ? (wins / total) * 100 : 0;
  }
}
